django/template/base.py

```diff
@@ -288,6 +294,59 @@ def get_exception_info(self, exception, token):
         }
 
 
+class PartialTemplate:
+    """A lightweight Template lookalike used for template partials.
+
+    Wraps nodelist as a partial, in order to be able to bind context.
+    """
+
+    def __init__(self, nodelist, origin, name):
+        self.nodelist = nodelist
+        self.origin = origin
+        self.name = name
+
+    def get_exception_info(self, exception, token):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return template.get_exception_info(exception, token)
+
+    def find_partial_source(self, full_source, partial_name):
+        start_match = None
+        nesting = 0
+
+        for match in combined_partial_re.finditer(full_source):
+            if match.group(1):  # Start tag
+                tag_name = match.group("name")
+                if tag_name == partial_name and start_match is None:
+                    start_match = match
+                    nesting = 1
+                elif start_match is not None:
+                    nesting += 1
+            else:  # End tag (group 2)
+                if start_match is not None:
+                    nesting -= 1
+                    if nesting == 0:
+                        return full_source[start_match.start() : match.end()]
```

Comment:
This can be a little less verbose:

```suggestion
        for match in combined_partial_re.finditer(full_source):
            if match["name"]:  # Start Tag
                if start_match is None and match["name"] == partial_name:
                    start_match = match
                if start_match is not None:
                    nesting += 1
            elif start_match is not None:
                nesting -= 1
                if nesting == 0:
                    return full_source[start_match.start() : match.end()]
```